docs/external-interop/origin_avalanche/validator_set_registry.md

```diff
@@ -0,0 +1,110 @@
+```solidity
+struct Validator {
+    bytes blsPublicKey;
+    uint64 weight;
+}
+
+struct ValidatorSet {
+    bytes32 avalancheBlockchainID;
+    Validator[] validators;
+    uint64 totalWeight;
+    uint64 pChainHeight;
+    uint64 pChainTimeStamp;
+}
+```
+A map from `avalancheBlockchainID` to the current validator set will be maintained in the contract. This blockchain ID
+may belong to the P-chain or any Avalanche L1. We do not need to keep any validator set of a given blockchain ID other
+than the most current as __this contract assumes that a quorum of signatures is always possible to be acquired for the
+most current validator set__. How this assumption is ensured is not the concern of this contract. 
+
+One special validator set is used to initialize this contract: The P-chain validator set which is current at the time
+of contract instantiation. This will be used as the _root of trust_ as discussed in the [overview](summary.md).
+
+A challenge in passing these validator sets to the registry contract is that they may constitute a data payload in excess
+of blockchain transaction limits. As such, it may be necessary to break up the serialized validator set data across
+ multiple transactions. This will certainly be the case when initializing the contract with a P-chain validator set.
+
+Each of these transaction should consist of two pieces of data to pass into the registry contract:
+An ICM message and a shard of the serialized data set. The ICM message should contain a validator set payload given
+by the following Solidity data type:
+```solidity
+struct ValidatorSetStatePayload {
+    bytes32 avalancheBlockchainID;
+    uint64 pChainHeight;
+    uint64 pChainTimestamp;
+    bytes32 validatorSetHash;
+    uint64 shardNumber;
+}
+```
+This payload should be identical for each shard, except for the `shardNumber` field. The contained data identifies which validator
+set we are updating. The `validatorSetHash` acts as a commitment to full validator set we intend to communicate. The 
+contract will need to buffer the shards internally until the hash of all shards (sorted by `shardNumber`) matches the 
+`validatorSetHash` field. The data shards are simply an array of bytes. Each of these transactions should be signed by 
+the current validator set of the chain identified by `avalancheBlockchainID` (if updating the set) or by the current 
+P-chain validator set (if registering the L1 for the first time).
+
+This scheme allows the sharding of update sets, which will be necessary for initializing the contract, but may not be
+necessary otherwise. Many L1s will not have large validator sets in the first place; the main concern is updating the
+P-chain. Once a validator set is registered, the data shards above can be switched to diffs, which we expect to be 
+significantly smaller. That is to say, we can specify that the register function expects data shards to be a raw 
+validator set, but the update function expects them to be diffs.
```

Comment:
I think the update function should be able to accept diffs _**or**_ a full set. The problem with diffs is that it's probably cheap to update the validator set on avalanche, but more expensive to update the set on ethereum. This could lead to a DoS vector.